### PR TITLE
Update loading animation

### LINKED ISSUE #6463: Improve default loading animation
**Description**

I believe the default loading animation should be improved as it is quite pixelated on certain types of screens. See https://github.com/eclipse-theia/theia/issues/4102#issuecomment-465586438. The same is true for Firefox.

<div align='center'>

image

</div>

**Additional Information**

A previous attempt was made in the past to update the animation (https://github.com/eclipse-theia/theia/pull/3156), but a more generic one should be present so it can be used by adopters without requiring the need to update it.

### PR DESCRIPTION
**Pull request by sh7dm**
#### What it does

Fixes #6463.

#### How to test

Visually, inspect SVG file of the image.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/style/variables-bright.useable.css [hunk 1]
```diff
@@ -33,10 +33,12 @@ is not optimized for dense, information rich UIs.
 
   --theia-ui-font-scale-factor: 1.2;
   --theia-ui-font-size0: calc(var(--theia-ui-font-size1) / var(--theia-ui-font-scale-factor));
-  --theia-ui-font-size1: 13px; /* Base font size */
```
Comment (vince-fugnitto) [inline packages/core/src/browser/style/variables-bright.useable.css:36]: I believe we should revert this change, the comment is directed specifically for the variable:
```css
--theia-ui-font-size1: 13px; /* Base font size */
```

File: packages/core/src/browser/style/variables-bright.useable.css [hunk 2]
```diff
@@ -33,10 +33,12 @@ is not optimized for dense, information rich UIs.
...
+  --theia-ui-font-size1: 13px;
+  /* Base font size */
   --theia-ui-font-size2: calc(var(--theia-ui-font-size1) * var(--theia-ui-font-scale-factor));
   --theia-ui-font-size3: calc(var(--theia-ui-font-size2) * var(--theia-ui-font-scale-factor));
-  --theia-ui-icon-font-size: 14px; /* Ensures px perfect FontAwesome icons */
```
Comment (vince-fugnitto) [inline packages/core/src/browser/style/variables-bright.useable.css:39]: Same as previous comment, the change should be reverted since it applies for the variable itself.

File: packages/core/src/browser/icons/spinner.svg [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<!--Licensed under the MIT License. See License.txt in the project root for license information.-->
```
Comment (vince-fugnitto) [inline packages/core/src/browser/icons/spinner.svg:1]: For my own reference, where does the image come from?
For example:
- what is the original link to the image?
- was the image created manually by you?

Comment (sh7dm) [reply packages/core/src/browser/icons/spinner.svg:1]: Idea from [this codepen](https://codepen.io/aleksander351/pen/KzgKPo), implemented myself.

**Comment by sh7dm** [comment]
@vince-fugnitto I fixed all requested changes. The contrast way the thing I thought about, but it seems to me that this color is great for almost any adequate background:
image
image
image
Probably we can make it a bit darker.
The another idea is to use foreground color for it, so it'd be perfect for anything, even with blue background.

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto I fixed all requested changes. The contrast way the thing I thought about, but it seems to me that this color is great for almost any adequate background:
> image
> image
> image
> Probably we can make it a bit darker.
> The another idea is to use foreground color for it, so it'd be perfect for anything, even with blue background.

It's not about visibility (contrast) but more consistency, a bright blue color in an otherwise all red theme does not look consistent with the rest of the application.

**Comment by sh7dm** [comment]
What do you think about using foreground color? SVG can be buitt into the page and use `currentColor` or any other of CSS powers.

**Comment by sh7dm** [comment]
Also, inlining the animation into page can remove blank screen while image is loading on slow network

**Comment by vince-fugnitto** [comment]
> What do you think about using foreground color? SVG can be buitt into the page and use `currentColor` or any other of CSS powers.

I think it's reasonable to use some of the exposed theming colors in order to properly style the loading screen. Whichever approach we do, I think it should remain possible for extenders/application developers to be able to replace the loading with their own custom ones.

**Comment by sh7dm** [comment]
May that be outside of the context of this PR? I can implement only theming with foreground color. I'm not very familiar with internal structure yet

**Comment by vince-fugnitto** [comment]
> May that be outside of the context of this PR? I can implement only theming with foreground color. I'm not very familiar with internal structure yet

I meant that when the pull-request is ready for another round of review, I'll also verify that it still holds true that extenders/application-developers are able to replace the loading animation with their own. We currently have applications built using theia, and I'd like to avoid breaking them.

**Comment by sh7dm** [comment]
Ah, okay. Apps can replace this at the moment? If so, inlining won't work

**Comment by vince-fugnitto** [comment]
> Ah, okay. Apps can replace this at the moment? If so, inlining won't work

Yes, theia is a framework so the image you are currently updating is the base one or default.
Applications have the ability to brand themselves and replace the loading animation with their own. The thing is, they do this by providing a custom preload html template which replaces the current one so it should still be possible if the image is inline (given that they somehow remove it using css). I'll create an example from which you can see, and also test against it.

**Comment by vince-fugnitto** [comment]
> What's about a plain splash screen, without animation, defaulting to Theia logo?

I think we used a default loading animation since it can be re-used by downstream applications, and also better expresses that loading is in progress. If we were to use the logo for a splash screen, I can see the following two issues:
- It forces downstream products to update the logo (since they cannot rely on the loading animation anymore). It will be fine for applications that already have custom loaders.
- It does not convey loading as much as an animation would.

This is however my opinion and others might have different insights. 
Is there a reason to not have the animation and only a logo?

**Comment by vince-fugnitto** [comment]
Just as an FYI, we had a previous loading animation WIP where we decided to keep the branding aspects separate: https://github.com/eclipse-theia/theia/pull/3156

**Comment by sh7dm** [comment]
okay, so probably to preserve compatibility we can do this:
1. Search for `.svg` loader in user's project
2. Search for `.gif` loader in user's project
3. Use `.svg` from Theia

**Comment by vince-fugnitto** [comment]
> okay, so probably to preserve compatibility we can do this:
> 
> 1. Search for `.svg` loader in user's project
> 2. Search for `.gif` loader in user's project
> 3. Use `.svg` from Theia

I'm not sure why we need all three steps, the first should be fine for adopters, and theia should keep a default loading animation in the framework as is today.

**Comment by sh7dm** [comment]
Hmm, do existing IDE's define their load as SVG? or do they provide their own preload html code?

**Comment by vince-fugnitto** [comment]
> Hmm, do existing IDE's define their load as SVG? or do they provide their own preload html code?

I'm not sure how others do it, but for an existing product I contribute to, the idea is that we have a preload html which includes an inline svg, and adds some css styling to remove the existing loading animation (so it doesn't appear momentarily on startup).

**Comment by sh7dm** [comment]
If it is the only way to customize it, we can go directly to step 3, user will have to remove it manually as before. In this case, if selectors remain the same, it'll be backwards-compatible.

**Comment by vince-fugnitto** [comment]
> If it is the only way to customize it, we can go directly to step 3, user will have to remove it manually as before. In this case, if selectors remain the same, it'll be backwards-compatible.

It should work, that's why I mentioned that during the review I'll test with a custom application if replacing the loading animation still works correctly.

**Comment by sh7dm** [comment]
@vince-fugnitto may I start working on inline SVG now?

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto may I start working on inline SVG now?

Yes, I wasn't trying to hold you back 😄

**Comment by sh7dm** [comment]
@vince-fugnitto done! The only issue is flickering, caused by #6999 . When it'll get fixed, loading will have good color every time.

File: packages/mini-browser/src/browser/mini-browser-content.ts [hunk 4]
```diff
@@ -297,6 +297,11 @@ export class MiniBrowserContent extends BaseWidget {
 
         const loadIndicator = document.createElement('div');
         loadIndicator.classList.add(MiniBrowserContentStyle.PRE_LOAD);
+        loadIndicator.innerHTML = `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="100px" height="100px" viewBox="0 0 100 100">
+  <circle cx="50" cy="50" fill="none" stroke="var(--theia-foreground, #949494)" stroke-width="10" r="35" stroke-dasharray="160 59.91">
+    <animateTransform attributeName="transform" type="rotate" repeatCount="indefinite" dur="1s" values="0 50 50;360 50 50" keyTimes="0;1"></animateTransform>
+  </circle>
+</svg>`;
```
Comment (vince-fugnitto) [inline packages/mini-browser/src/browser/mini-browser-content.ts:304]: I don't like that we have to duplicate the code for the spinner. In fact, we should be able to use `codicons` for this spinner:
- https://code.visualstudio.com/api/references/icons-in-labels#animation

File: packages/core/src/browser/icons/spinner.svg (hunk 3 above)
Comment (vince-fugnitto) [inline packages/core/src/browser/icons/spinner.svg:1]: Any reason for using MIT as a license? The link "see license.txt" is bogus since no MIT license exists.

Comment (sh7dm) [reply packages/core/src/browser/icons/spinner.svg:1]: All files have this header. Okay, will remove.

File: packages/core/src/browser/style/variables-bright.useable.css [hunk 5]
```diff
@@ -61,7 +61,6 @@ is not optimized for dense, information rich UIs.
   --theia-icon-arrow-down: url(../icons/arrow-down-bright.svg);
   --theia-sprite-y-offset: 0px;
   --theia-icon-circle: url(../icons/circle-bright.svg);
-  --theia-preloader: url(../icons/spinner.gif);
```
[the remark this answers is not in the document]
Comment (sh7dm) [reply packages/core/src/browser/style/variables-bright.useable.css:64]: Will restore.

File: packages/core/src/browser/icons/spinner.svg (hunk 3 above)
Comment (vince-fugnitto) [reply packages/core/src/browser/icons/spinner.svg:1]: > All files have this header. Okay, will remove.

Which files exactly?

File: packages/mini-browser/src/browser/mini-browser-content.ts (hunk 4 above)
Comment (sh7dm) [reply packages/mini-browser/src/browser/mini-browser-content.ts:304]: Hmm, codicon is a good idea. Will try

Comment (sh7dm) [reply packages/mini-browser/src/browser/mini-browser-content.ts:304]: This syntax can only be used in labels, but we can use other icons easily.

Comment (sh7dm) [reply packages/mini-browser/src/browser/mini-browser-content.ts:304]: Did it! Which one do you like more, `ease` or `linear`? `ease` looks better for me.
Peek 2020-04-08 20-48
Peek 2020-04-08 20-47

File: packages/core/src/browser/icons/spinner.svg (hunk 3 above)
Comment (sh7dm) [reply packages/core/src/browser/icons/spinner.svg]: Another icon. I looked there to know how are they formatted and what root props they have.

File: dev-packages/application-manager/src/generator/frontend-generator.ts [hunk 6]
```diff
@@ -53,7 +53,10 @@ export class FrontendGenerator extends AbstractGenerator {
 </head>
 
 <body>
-  <div class="theia-preload">${this.compileIndexPreload(frontendModules)}</div>
+  <div class="theia-preload">
+    ${this.compileIndexPreload(frontendModules)}
+    <i class="codicon codicon-sync theia-loader"></i>
```
Comment (akosyakov) [inline dev-packages/application-manager/src/generator/frontend-generator.ts:58]: It will break all clients having custom index preload?

Comment (sh7dm) [reply dev-packages/application-manager/src/generator/frontend-generator.ts:58]: @akosyakov how can we avoid this?

Comment (sh7dm) [reply dev-packages/application-manager/src/generator/frontend-generator.ts:58]: Probably we can use a pseudo-element of `theia-preload`, but we'll have to duplicate codicon's style setting its background. Will try to do so.

Comment (sh7dm) [reply dev-packages/application-manager/src/generator/frontend-generator.ts]: Works fine, the only problem is that bad character is displayed a bit of time while font is being loaded.

Comment (akosyakov) [reply dev-packages/application-manager/src/generator/frontend-generator.ts]: There is a code above which returns if the template is missing. So far it was returning an empty string, you could return something else there. I am not sure that adding/changing styles is good.

Comment (sh7dm) [reply dev-packages/application-manager/src/generator/frontend-generator.ts]: Which style exactly? Where's that code?

File: dev-packages/application-manager/src/generator/frontend-generator.ts [hunk 7]
```diff
@@ -53,7 +53,7 @@ export class FrontendGenerator extends AbstractGenerator {
 </head>
 
 <body>
-  <div class="theia-preload">${this.compileIndexPreload(frontendModules)}</div>
+  <div class="theia-preload theia-loader">${this.compileIndexPreload(frontendModules)}</div>
```
Comment (vince-fugnitto) [inline dev-packages/application-manager/src/generator/frontend-generator.ts:56]: I believe the following approach breaks downstream applications which use `theia-preload` to define their own custom spinners. At the moment, they may provide their own loading animation or splash screen with `compileIndexPreload` and the `theia-loader` styling would need to be overwritten.

**Review by vince-fugnitto — COMMENTED** [review]
@sh7dm do you still plan on updating the pull-request, the current approach seems to have issues for downstream apps.

**Comment by sh7dm** [comment]
Not sure how to avoid that breakage. Do you have any examples of code from projects overriding that?

**Comment by vince-fugnitto** [comment]
> Not sure how to avoid that breakage. Do you have any examples of code from projects overriding that?

The breakage was previously discussed during the following [review](https://github.com/eclipse-theia/theia/pull/7513#discussion_r405999353). Essentially, thanks to #4232, it is possible for downstream applications to define custom preload elements in order to customize the loading screen to suit their needs. With these changes, such applications are broken because they do not expect the new stylings including the`codicon-refresh` spinner.

The following is an application which defines in its `package.json` a configuration for setting the [`preload`](https://github.com/vince-fugnitto/theia-custom-preload/blob/dfbc41479d9f50d9b91dde19c540bba17a81e719/package.json#L12-L16):

```json
"generator": {
    "config": {
        "preloadTemplate": "./preload.html"
    }
}
```

The preload can be anything an application needs for their use-case, in the example it is simply a [spinner](https://github.com/vince-fugnitto/theia-custom-preload/blob/main/preload.html#L1-L56) which overrides the default one.

The contribution is provided by the following default props:

https://github.com/eclipse-theia/theia/blob/8d5e12accfbd6bbe0cd720d3ebd151b700f4dc80/dev-packages/application-package/src/application-props.ts#L75-L95

File: dev-packages/application-manager/src/generator/frontend-generator.ts (hunk 7 above)
Comment (sh7dm) [reply dev-packages/application-manager/src/generator/frontend-generator.ts:56]: I have no idea how do we make a css style just disappear instead of overriding it when somebody defines `.theia-preload`

Comment (vince-fugnitto) [reply dev-packages/application-manager/src/generator/frontend-generator.ts:56]: I don’t think anyone is asking for stylings to disappear, only that the new classname and introduction of the codicon-refresh spinner break downstream applications unnecessarily (ex: who have introduced their own preload template). 

The following may be more aligned with what the changes should be: https://github.com/eclipse-theia/theia/pull/7513#discussion_r406039856

**Comment by sh7dm** [comment]
Alright, a good solution would be returning a div with theia-loader instead of an empty string from `compileIndexPreload`.

**Comment by vince-fugnitto** [comment]
@JonasHelming I think the changes are no longer necessary, applications can easily add their custom loading or splash screens similarly to https://github.com/eclipse-theia/theia-blueprint/pull/73. I don't think that applications will keep the default spinner for long. Do you think we should close?
